### ISSUE TITLE
breadcrumbs not appearing for optional route paramters

### ISSUE THREAD

**Issue by zulfadlirahmat** [maartenba/MvcSiteMapProvider#387]
Hi,

Currently i have this node set up in my Mvc.sitemap.

```xml
<mvcSiteMapNode title="Add on Services" controller="AddOnServices" action="Index">
    <mvcSiteMapNode title="Flight" controller="AddOnServices" action="Flight"></mvcSiteMapNode>
</mvcSiteMapNode>
```

So when i hit the URL, http://localhost/addonservices/flight

The breadcrumbs will appear as HOME > Add on Services > Flight

But when i have optional parameters in the URL, http://localhost/addonservices/flight/1/2.

The routeconfig are currently set as:
```c#
routes.MapRoute(
                name: "addonservices.flight",
                url: "addonservices/flight/{airline}/{type}",
                defaults: new { controller = "AddOnServices", action = "Flight", airline = UrlParameter.Optional , type = UrlParameter.Optional }
            );
```

The breadcrumbs disappears. I have tried adding in preservedRouteAttributes for both optional values but to no avail.

Please advise on how i can achieve breadcrumbs for both scenarios.

Thank you.

**Comment by NightOwl888** [comment]
The route that you have provided does not match the second URL that you provided (`http://localhost/addonservices/flight/1/2`). You must be matching a different route in this case.

Whether the route parameters are *optional* or not does not matter. You must account for every route parameter that is not `area`, `controller` or `action`. You should supply them whether they are optional or required when using `preservedRouteParameters`.

Unfortunately, because you have not supplied a matching route I can't be specific. But I can tell you that whatever the key names of `1` and `2` are in the URL `http://localhost/addonservices/flight/1/2` should be put into `preservedRouteParameters`.

If you don't have a matching route for that URL, then that is a problem with your MVC route configuration. You should have a route that matches the signature of the URL or MVC won't be able to match it.

```c#
	routes.MapRoute(
		name: "addonservices.flight.option1.option2",
		url: "addonservices/flight/{option1}/{option2}",
		defaults: new { controller = "AddOnServices", action = "Flight" }
	);

	routes.MapRoute(
		name: "addonservices.flight",
		url: "addonservices/flight/{airline}",
		defaults: new { controller = "AddOnServices", action = "Flight", airline = UrlParameter.Optional }
	);
```

And then to match the same URL with MvcSiteMapProvider, you need to specify the same values in the preservedRouteParameters of the node.

```xml
<mvcSiteMapNode title="Add on Services" controller="AddOnServices" action="Index">
    <mvcSiteMapNode title="Flight" controller="AddOnServices" action="Flight" preservedRouteParamters="option1,option2"></mvcSiteMapNode>
</mvcSiteMapNode>
```

**Comment by zulfadlirahmat** [comment]
Hi NightOwl888,

Thank you for your reply. I have edited the routes. Pasted the wrong the route values that i have in my application.

I did exactly what you said, using the preservedRouteParameters to ensure that the parameters are similiar to what they have been set in the RouteConfig file.

It is still not appearing. Does having dynamicnodes for some of the nodes causes an issue with the preservedRouteParameters ?

<pre>
<mvcSiteMapNode title="Add on Services" controller="AddOnServices" action="Index">
	<mvcSiteMapNode title="Flight" controller="AddOnServices" action="Flight" preservedRouteParamters="airline,type">
		<mvcSiteMapNode title="Details" controller="AddOnServices" action="FlightDetails" dynamicNodeProvider="Tale.SS.MVCSiteMapNodeProvider.FlightDynamicNodeProvider, Tale.SS" />
	</mvcSiteMapNode>
</mvcSiteMapNode>
</pre>

If it helps, the action looks like this

<pre>
 public ActionResult Flight(int? page, int? airline, string type, string sort_order)
{

}
</pre>

So basically on first entry, the breadcrumbs will appear(http://localhost/addonservices/flight/).

 But if there are parameters present in the URL(http://localhost/addonservices/flight/1/sq), then the there is a mismatch? Am i configuring it wrong ?

**Comment by NightOwl888** [comment]
> Interestingly, if i were to hit the URL this way (http://localhost/addonservices/flight?airline=1&type=sq the breadcrumbs appear. Could this be due to the URL rewrite then?

Definitely. You didn't mention before you are using URL rewriting, but it is very unusual to use it in conjunction with routing because routing supports 2 way URL mapping and URL rewriting does not. When using routing you can *generate* a URL based on the route in addition to matching the incoming route. With URL rewriting you cannot generate URLs.

Typically, you would add as many routes as you need with required and/or optional parameters to match all of the segment lengths that you require. When you use routes the URL parameters are copied into the route value collection. However, when you use query string parameters the parameters are not copied into the route value collection. In that case, they are ignored by MvcSiteMapProvider by default and only utilized if you explicitly configure them as values or preservedRouteParameters.

I have never tested MvcSiteMapProvider with URL rewriting and I suspect it won't work right if you use URL rewriting.

**Comment by zulfadlirahmat** [comment]
Hmmm,

Thanks for the reply Nightowl88. Im not sure if  the term to use is URL re-writing. What i meant to say is for instance is that without configuring the Routeconfig to have this routes:

<pre>
routes.MapRoute(
                name: "addonservices.flight",
                url: "addonservices/flight/{airline}",
                defaults: new { controller = "AddOnServices", action = "Flight", airline = UrlParameter.Optional, type = UrlParameter.Optional }
            );

routes.MapRoute(
                name: "addonservices.flightnettratedetails",
                url: "addonservices/flight/details/{_flightid}",
                defaults: new { controller = "AddOnServices", action = "FlightDetails" }
				constraints: new { _flightid = @"\d+" }
            );
</pre>

Route "addonservices.flight" will generate (http://localhost/addonservices/flight/{airline}) while route "addonservices.flightnettratedetails" wille generate http://localhost/addonservices/flightdetails/{_flightid}). As i want the url to produce the following, http://localhost/flight/details, i would need to explicitly configure routeconfig to have the exact URL i want for easy identification.

Is this called URL rewriting via the use of RouteConfig ?

What i dont understand is how come http://localhost/addonservices/flight would generate the breadcrumbs while http://localhost/addonservices/flight/1/sq doesnt match the MvcSiteMapProvider configurations even with the use of preservedRouteParameters.

I changed how mvc display the URL via the Routeconfig for all the modules, could this cause MvcSiteMapProvider to be confused even if i have indicated the exact controller and action that the route mappings uses ?

**Comment by NightOwl888** [comment]
A URL like `http://localhost/addonservices/flight/1/sq` will not match any of the MVC routes you have provided thus far. This is the problem, it has nothing to do with MvcSiteMapProvider.

MvcSiteMapProvider works off of the route values dictionary and cannot match this information unless you supply it there. You need a route to match the entire URL and put those extra values into the route values dictionary. I gave you an example before:

```c#
	routes.MapRoute(
		name: "addonservices.flight.option1.option2",
		url: "addonservices/flight/{option1}/{option2}",
		defaults: new { controller = "AddOnServices", action = "Flight" }
	);
```

But only you know what `1` and `sq` actually mean, so you need to put them into route values that are named in a meaningful way for your scenario. Then, you can use those same names in MvcSiteMapProvider to force a match with `preservedRouteParameters`.

But again, the crux of your problem is that you don't have a valid route that this URL (`http://localhost/addonservices/flight/1/sq`) will match (or at least you have not supplied one here). It may be matching the URL `addonservices/flight/{airline}` and then incorrectly putting the `1` into the `airline` key and then simply tossing out the `sq`, but I have not tried it to see what happens in this case. The bottom line is you are not accounting for this extra data so the routing framework is ignoring it or misplacing it instead of putting it where it needs to be.

> Is this called URL rewriting via the use of RouteConfig ?

No, this is routing. You used the term URL rewriting incorrectly.

> I changed how mvc display the URL via the Routeconfig for all the modules, could this cause MvcSiteMapProvider to be confused even if i have indicated the exact controller and action that the route mappings uses ?

Possibly. But, you need to account for *all of the route values* (not just controller and action) for each route. 

MvcSiteMapProvider works off of the principle that the route value combination for each node *must* be unique. This implicitly means that all route values need to be accounted for both in your routing configuration and in the MvcSiteMapProvider node configuration in order to get a match. You will only get a match when the values on both sides (the route values dictionary and the MvcSiteMapProvider route values dictionary) are the same.

**Comment by NightOwl888** [comment]
I am closing this issue as it was caused by a misconfiguration of .NET routing, and is now presumed to be resolved.
